### ISSUE TITLE
manual_swap suggests invalid multiple mutable borrow

### ISSUE THREAD

**Issue by Luminoth** [rust-lang/rust-clippy#8154]
### Description

Given the following function:

```
fn swap_xy(mut v: IVec3) -> IVec3 {
    let x = v.x;
    v.x = v.y;
    v.y = x;

    v
}
```

Clippy is suggesting the following:

```
warning: this looks like you are swapping `v.x` and `v.y` manually
  --> day19/src/main.rs:15:5
   |
15 | /     let x = v.x;
16 | |     v.x = v.y;
17 | |     v.y = x;
   | |___________^ help: try: `std::mem::swap(&mut v.x, &mut v.y)`
   |
   = note: `#[warn(clippy::manual_swap)]` on by default
   = note: or maybe you should use `std::mem::replace`?
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_swap
```

which produces the following error:

```
error[E0499]: cannot borrow `v` as mutable more than once at a time
  --> day19/src/main.rs:15:35
   |
15 |     std::mem::swap(&mut v.x, &mut v.y);
   |     --------------      -         ^ second mutable borrow occurs here
   |     |                   |
   |     |                   first mutable borrow occurs here
   |     first borrow later used by call
```

It looks like this may have been fixed in https://github.com/rust-lang/rust-clippy/issues/981 and then maybe reintroduced?

### Version

```text
rustc 1.57.0 (f1edd0429 2021-11-29)
binary: rustc
commit-hash: f1edd0429582dd29cccacaf50fd134b05593bd9c
commit-date: 2021-11-29
host: x86_64-unknown-linux-gnu
release: 1.57.0
LLVM version: 13.0.0
```

**Comment by Jarcho** [comment]
There's something wrong with your example. This compiles fine

```rust
struct IVec3 {
    x: u32,
    y: u32
}

fn swap_xy(mut v: IVec3) -> IVec3 {
    std::mem::swap(&mut v.x, &mut v.y);
    v
}
```

**Comment by Luminoth** [comment]
I'm not sure what your point is. I did what the clippy message suggested and it didn't compile. The message shouldn't be giving me a suggestion that doesn't build.

**Comment by Luminoth** [comment]
Ah I see what my example is missing. Here's the full thing:

```
use glam::IVec3;

fn swap_xy(mut v: IVec3) -> IVec3 {
    std::mem::swap(&mut v.x, &mut v.y);

    v
}
```

So it's giving an invalid suggestion, but only when using the glam vector.

**Comment by Jarcho** [comment]
Looks like glam gets it's fields from a deref impl, which does cause the issue. Minimal example would be:

```rust
struct Foo {
    x: u32,
    y: u32,
}
struct Bar(Foo);
impl core::ops::Deref for Bar {
    type Target = Foo;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
impl core::ops::DerefMut for Bar {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

fn swap(x: &mut Bar) {
    let t = x.x;
    x.x = x.y;
    x.y = t;
}
```
